src/proxy/pool.rs

```diff
@@ -12,241 +12,1413 @@
+        );
+        let after_dropcount = conn_drop_counter.load(Ordering::Relaxed);
+        assert!(
+            after_dropcount == 4,
+            "after dropcount was {after_dropcount}"
+        );
+        server_drain_signal.drain().await;
+        server_handle.await.unwrap();
+
+        drop(pool);
+    }
+
+    fn spawn_client(
+        mut pool: WorkloadHBONEPool,
+        key: WorkloadKey,
+        remote_addr: SocketAddr,
+        req_count: u32,
+    ) -> task::JoinHandle<()> {
+        tokio::spawn(async move {
+            let req = || {
+                hyper::Request::builder()
+                    .uri(format!("{remote_addr}"))
+                    .method(hyper::Method::CONNECT)
+                    .version(hyper::Version::HTTP_2)
+                    .body(Empty::<Bytes>::new())
+                    .unwrap()
+            };
+
+            let start = Instant::now();
+
+            let mut c1 = pool
+                .connect(key.clone())
+                // needs tokio_unstable, but useful
+                // .instrument(tracing::debug_span!("client_tid", tid=%tokio::task::id()))
+                .await
+                .unwrap();
+            debug!(
+                "client spent {}ms waiting for conn",
+                start.elapsed().as_millis()
+            );
+
+            let mut count = 0u32;
+            loop {
+                count += 1;
+                let res = c1.send_request(req()).await;
+
+                if res.is_err() {
+                    panic!("SEND ERR: {:#?} sendcount {count}", res);
+                } else if res.unwrap().status() != 200 {
+                    panic!("CLIENT RETURNED ERROR")
+                }
+
+                if count >= req_count {
+                    debug!("CLIENT DONE");
+                    break;
+                }
+            }
+        })
+    }
+
+    fn spawn_persistent_client(
+        mut pool: WorkloadHBONEPool,
+        key: WorkloadKey,
+        remote_addr: SocketAddr,
+        stop: Watch,
+    ) -> task::JoinHandle<()> {
+        tokio::spawn(async move {
+            let req = || {
+                hyper::Request::builder()
+                    .uri(format!("{remote_addr}"))
+                    .method(hyper::Method::CONNECT)
+                    .version(hyper::Version::HTTP_2)
+                    .body(Empty::<Bytes>::new())
+                    .unwrap()
+            };
+
+            let start = Instant::now();
+
+            let mut c1 = pool
+                .connect(key.clone())
+                // needs tokio_unstable, but useful
+                // .instrument(tracing::debug_span!("client_tid", tid=%tokio::task::id()))
+                .await
+                .unwrap();
+            debug!(
+                "client spent {}ms waiting for conn",
+                start.elapsed().as_millis()
+            );
+
+            let mut count = 0u32;
+            // send forever, once we get a conn, until someone signals us to stop
```

Comment:
Probably don't need to infinite loop on sending requests, but rather have one request we keep open forever

Reply:
Yep, this was just making some tests more confusing/flaky anyway.